File: src/audio.rs
#[cfg(feature = "alsa")]
mod imp {
    use std::sync::Mutex;

    const SAMPLE_RATE: u32 = 44100;
    const CHANNELS: u32 = 2;
    const CLICK_DURATION_MS: u32 = 30;
    const CLICK_FREQ: f32 = 800.0;
    const CLICK_VOLUME: f32 = 0.4;

    struct NavAudio {
        pcm: alsa::pcm::PCM,
        click_buffer: Vec<i16>,
    }

    static AUDIO: Mutex<Option<NavAudio>> = Mutex::new(None);

    fn generate_click() -> Vec<i16> {
        let num_samples = (SAMPLE_RATE * CLICK_DURATION_MS / 1000) as usize;
        let mut buf = Vec::with_capacity(num_samples * CHANNELS as usize);

        for i in 0..num_samples {
            let t = i as f32 / SAMPLE_RATE as f32;
            let mut envelope = 1.0 - (i as f32 / num_samples as f32);
            envelope *= envelope;
            let sample =
                (2.0 * std::f32::consts::PI * CLICK_FREQ * t).sin() * CLICK_VOLUME * envelope;
            let s = (sample * 32767.0) as i16;
            buf.push(s);
            buf.push(s);
        }
        buf
    }

    pub fn audio_init() {
        let pcm = match alsa::pcm::PCM::new("default", alsa::Direction::Playback, false) {
            Ok(p) => p,
            Err(e) => {
                eprintln!("audio init error: {e}");
                return;
            }
        };

        {
            let hwp = match alsa::pcm::HwParams::any(&pcm) {
                Ok(h) => h,
                Err(e) => {
                    eprintln!("audio init hw_params error: {e}");
                    return;
                }
            };
            if let Err(e) = hwp.set_channels(CHANNELS) {
                eprintln!("audio init channels error: {e}");
                return;
            }
            if let Err(e) = hwp.set_rate(SAMPLE_RATE, alsa::ValueOr::Nearest) {
                eprintln!("audio init rate error: {e}");
                return;
            }
            if let Err(e) = hwp.set_format(alsa::pcm::Format::s16()) {
                eprintln!("audio init format error: {e}");
                return;
            }
            if let Err(e) = hwp.set_access(alsa::pcm::Access::RWInterleaved) {
                eprintln!("audio init access error: {e}");
                return;
            }
            if let Err(e) = pcm.hw_params(&hwp) {
                eprintln!("audio init set hw_params error: {e}");
                return;
            }
        }

        let nav = NavAudio {
            pcm,
            click_buffer: generate_click(),
        };

        *AUDIO.lock().unwrap() = Some(nav);
        eprintln!("audio: initialized");
    }

    pub fn audio_play() {
        let mut lock = AUDIO.lock().unwrap();
        if let Some(ref mut audio) = *lock {
            let _ = audio.pcm.prepare();
            let io = audio.pcm.io_i16();
            if let Ok(io) = io {
                let _ = io.writei(&audio.click_buffer);
            }
        }
    }
}

#[cfg(feature = "alsa")]
pub use imp::{audio_init, audio_play};

#[cfg(not(feature = "alsa"))]
pub fn audio_init() {
    eprintln!("audio: disabled (no ALSA)");
}

#[cfg(not(feature = "alsa"))]
pub fn audio_play() {}

File: src/audio.cpp
#include "audio.h"
#include <print>

#ifdef ARKANA_HAS_ALSA

#include <cmath>
#include <iostream>
#include <memory>
#include <mutex>
#include <numbers>
#include <vector>

#include "nothrow.h"

#include <alsa/asoundlib.h>

constexpr int SAMPLE_RATE = 44100;
constexpr int CHANNELS = 2;
constexpr int CLICK_DURATION_MS = 30;
constexpr float CLICK_FREQ = 800.0f;
constexpr float CLICK_VOLUME = 0.4f;

struct NavAudio {
    snd_pcm_t* pcm;
    std::vector<int16_t> click_buffer;
    std::mutex mutex;
};

static std::unique_ptr<NavAudio> g_audio;

static std::vector<int16_t> generate_click() {
    int num_samples = SAMPLE_RATE * CLICK_DURATION_MS / 1000;
    std::vector<int16_t> buf(num_samples * CHANNELS);

    for (int i = 0; i < num_samples; i++) {
        float t = static_cast<float>(i) / SAMPLE_RATE;
        float envelope = 1.0f - static_cast<float>(i) / num_samples;
        envelope = envelope * envelope;
        float sample = std::sin(2.0f * std::numbers::pi_v<float> * CLICK_FREQ * t) * CLICK_VOLUME * envelope;
        int16_t s = static_cast<int16_t>(sample * 32767.0f);
        buf[i * 2] = s;
        buf[i * 2 + 1] = s;
    }
    return buf;
}

void audio_init() {
    snd_pcm_t* pcm = nullptr;
    int err = snd_pcm_open(&pcm, "default", SND_PCM_STREAM_PLAYBACK, 0);
    if (err < 0) {
        std::println(std::cerr, "audio init error: {}", snd_strerror(err));
        return;
    }

    snd_pcm_set_params(pcm,
                       SND_PCM_FORMAT_S16_LE,
                       SND_PCM_ACCESS_RW_INTERLEAVED,
                       CHANNELS,
                       SAMPLE_RATE,
                       1,
                       100000);

    auto nav = make_nothrow<NavAudio>();
    nav->pcm = pcm;
    nav->click_buffer = generate_click();
    g_audio = std::move(nav);

    std::println(std::cerr, "audio: initialized");
}

void audio_play() {
    if (!g_audio) return;

    std::lock_guard<std::mutex> lock(g_audio->mutex);
    snd_pcm_t* pcm = g_audio->pcm;
    snd_pcm_prepare(pcm);
    const auto& buf = g_audio->click_buffer;
    snd_pcm_writei(pcm, buf.data(), buf.size() / CHANNELS);
}

#else

void audio_init() {
    std::println(std::cerr, "audio: disabled (no ALSA)");
}

void audio_play() {}

#endif
